### PR TITLE
Post deployment information

### PR DESCRIPTION
**Pull request by Lukewh**
Replaces https://github.com/juju/juju-gui/pull/3135

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8072/

**Comment by Lukewh** [comment]
## QA

- Create a charm or bundle, probably use one you have already, create a getstarted.md like https://gist.github.com/Lukewh/a8e4f04d1de7681ac05fcaa07b4facad and push your charm.
- Add your charm (with the getstarted.md) to the canvas.
- Deploy.
- See the panel.
- Celebrate.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8073/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8083/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8084/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8085/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8086/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8087/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8088/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8099/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8100/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8101/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8102/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8103/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8104/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8106/

**Comment by hatched** [comment]
Looks like legit test failures.

**Comment by hatched** [comment]
Is this ready for another review/qa?

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8113/

File: jujugui/static/gui/src/app/components/post-deployment/post-deployment.js [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+          </span>
+          <div onClick={this._handleContentClick.bind(this)}
+            dangerouslySetInnerHTML={{__html: this.state.content}} />
+        </Panel>
+      );
+    } else if (this.state.displayName) {
+      classes.push('post-deployment--simple');
+      return (
+        <Panel
+          extraClasses={classes.join(' ')}
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/post-deployment/post-deployment.js]: Could we remove this extra prop and just change the instanceName?

File: jujugui/static/gui/src/app/init/component-renderers-mixin.js [hunk 2]
```diff
@@ -342,6 +343,56 @@ const ComponentRenderersMixin = (superclass) => class extends superclass {
...
+    @param {Function} next Run the next handler.
+  */
+  _displayPostDeployment(state, next) {
+    if (!state.postDeploymentPanel.show) {
+      next();
+      return;
+    }
+    const entityId = state.postDeploymentPanel.entityId;
+    this.state.changeState({
+      'postDeploymentPanel': null
```
Comment (hatched) [inline jujugui/static/gui/src/app/init/component-renderers-mixin.js]: quotes aren't necessary here.

So from within the state callback for the post deployment you then want to clear it? Shouldn't it only be cleared when they hit X?

File: jujugui/static/gui/src/app/views/utils.js [hunk 3]
```diff
@@ -1613,6 +1613,11 @@ YUI.add('juju-view-utils', function(Y) {
         protocol: config.socket_protocol,
         uuid: model.uuid
       });
+      app.state.changeState({
+        'postDeploymentPanel': {
```
Comment (hatched) [inline jujugui/static/gui/src/app/views/utils.js]: quotes.

File: jujugui/static/gui/src/app/components/panel/panel.js [hunk 4]
```diff
@@ -25,6 +25,7 @@ class Panel extends React.Component {
     return classNames(
       'panel-component',
       this.props.instanceName,
+      this.props.extraClasses || null,
```
[the remark this answers is not in the document]
Comment (Lukewh) [reply jujugui/static/gui/src/app/components/panel/panel.js]: Yeah, the problem with that is to do the BEMs you need both the base class ('post-deployment' in this case) along with the modifier ('post-deployment--simple') so the whole class attr would be `class="post-deployment post-deployment--simple"`. Just having `class="post-deployment--simple"` wouldn't work.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8121/

File: jujugui/static/gui/src/app/components/entity-details/header/header.js [hunk 5]
```diff
@@ -47,6 +47,11 @@ class EntityHeader extends React.Component {
       var id = entity.id.replace('cs:', '');
       this.props.getBundleYAML(id, this._getBundleYAMLCallback.bind(this));
     }
+    this.props.changeState({
+      'postDeploymentPanel': {
```
[the remark this answers is not in the document]
Comment (Lukewh) [reply jujugui/static/gui/src/app/components/entity-details/header/header.js]: done

File: jujugui/static/gui/src/app/components/entity-details/header/header.js [hunk 6]
```diff
@@ -47,6 +47,11 @@ class EntityHeader extends React.Component {
       var id = entity.id.replace('cs:', '');
       this.props.getBundleYAML(id, this._getBundleYAMLCallback.bind(this));
     }
+    this.props.changeState({
```
[the remark this answers is not in the document]
Comment (Lukewh) [reply jujugui/static/gui/src/app/components/entity-details/header/header.js]: done

File: jujugui/static/gui/src/app/components/post-deployment/post-deployment.js [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+  /**
+    Callback handler for getting the entity details.
+
+    @param {Object} error Error from the API.
+    @param {Array} entityData The returned data.
+  */
+  _getEntityCallback(error, entityData) {
+    if (error) {
+      console.error(error);
+      console.error(`Entity not found with id: ${this.props.entityId}`);
```
[the remark this answers is not in the document]
Comment (Lukewh) [reply jujugui/static/gui/src/app/components/post-deployment/post-deployment.js]: yes, and done

File: jujugui/static/gui/src/app/components/post-deployment/post-deployment.js (hunk 1 above)
Comment (Lukewh) [reply jujugui/static/gui/src/app/components/post-deployment/post-deployment.js]: see above comment re. bem

File: jujugui/static/gui/src/app/init/component-renderers-mixin.js (hunk 2 above)
Comment (Lukewh) [reply jujugui/static/gui/src/app/init/component-renderers-mixin.js]: done

File: jujugui/static/gui/src/app/views/utils.js (hunk 3 above)
Comment (Lukewh) [reply jujugui/static/gui/src/app/views/utils.js]: done

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8122/

**Comment by huwshimi** [comment]
Deploying the model shows the charm panel, but clicking close on the panel shows: `Uncaught Error: React DOM tree root should always have a node reference.` in the console.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8126/

**Pull request merged**